Part 1 - Initial Review

whats-a-handle opened the pull request:
### Install instructions:

1. Clone repo `git@github.com:whats-a-handle/todo-list-app.git`
2. Change branch to `part-1`
3. Navigate to `frontend` folder and `npm install`
4. Run project with `npm start`


### Data Structure
Each ToDoItem is saved as an object like below. Each item is added as a key:value pair to the `todoItems` object in state within App.js. The `key` is the `itemName `and the value is the attributes of a ToDo Item. Each ToDo Item needs to have a unique `itemName` since this is used as the identifier. 
ToDoItem:
```
itemName: {
   itemName : String,
   isCompleted : Boolean,
   position: Number
}
```
### Functionality
- Add items to a Todo list
- Remove items from Todo list
- Update existing Todo items 
- Mark Todo Items as complete
- Reorder Todo Items via drag and drop via [react-dnd](https://github.com/react-dnd/react-dnd). 
    - Sometimes multiple items get moved when trying to move one item. Not sure if it's my code or the package. I noticed the issues board for this package has mentions of this occurring for other users as well. 

### Tricky stuff I ran into
- Had some difficulty tracking the position of various items, especially when they were deleted and rerendering the list. I have a working solution, but it feels hacky to me 🤔 . 
 - Whenever there's an update to the `todoItems` object in state, I generate an array from the values of the object and then do a sort on the position property for each TodoItem in that object. 
- State management and organization of where state should be kept and updated from "prop drilling?"
- Drag and Drop reordering with `react-dnd`. Followed this [guide here ](https://blog.logrocket.com/drag-and-drop-react-dnd/)and the docs, but I still dont really have a handle on the implementation. This was definitely a good deal of copy paste and some trial and error 😂 . 

### Browsers Tested
- Working on Chrome, Firefox, Edge, and Safari

### Main Packages used
- [react-dnd](https://github.com/react-dnd/react-dnd)
- [react material ui](https://github.com/mui/material-ui)

### Tests
- None

whats-a-handle committed:
Add initial commit of create-react-app frontend

whats-a-handle committed:
Messing with a ToDoListItem and ToDoList component

whats-a-handle committed:
Styling and centering todo list interface with Grid and Containers

whats-a-handle committed:
Add to list of todoitems (no key per item yet), super basic

whats-a-handle committed:
Using object to store todo items instead of array

whats-a-handle committed:
Added initial check-uncheck items

whats-a-handle committed:
Delete todo items from list

whats-a-handle committed:
prevent null or empty todo from being created

whats-a-handle committed:
Styling changes for gray background and white ui

whats-a-handle committed:
Rearrange items with state persistence...roooough

whats-a-handle committed:
Removed unnessary useState on todoItemOrderedList

whats-a-handle committed:
Initial work on inline editing on todo items

whats-a-handle committed:
Updated styling, added edit ability to TodoItems, added enter key to submit, renamed some properties of todoitem object

whats-a-handle committed:
simplified position calculation for order of items

whats-a-handle committed:
Strikethrough text on ToDoItems when isCompleted is true. Additionally, ToDoItems text will now wrap correctly

JWA111 commented on frontend/src/App.js:
```diff
@@ -0,0 +1,82 @@
+import CreateItemForm from './Components/CreateItemForm'
+import {Container,Grid} from '@mui/material';
+import './App.css';
+export default function App() {
```
I'd recommend Keeping your `App` component pretty simple as it is your root component and entry point. I recommend that you move the stuff specific to the todo list to the `TodoList` component instead.

JWA111 commented on frontend/src/App.js:
```diff
@@ -0,0 +1,82 @@
+import './App.css';
+export default function App() {
+  const [todoItems, setTodoItems] = useState({});
+  const [todoItemsOrderedList,setTodoItemsOrderedList] = useState([]);
```
Why manage two different copies of the data? I think it would be better to just have one copy of the data, probably as an already ordered list.

JWA111 commented on frontend/src/App.js:
```diff
@@ -0,0 +1,82 @@
+    const newToDo = {
+      itemName : itemName,
+      isCompleted : false,
+      position : 0, 
```
rather than have the block of conditional logic below, could set the position here:

```suggestion
      position : todoItems.length,  // assuming todoItems is an array
```

JWA111 replied:
Might also be good to add a unique `id` field, spoilers for the upcoming database part, but an ID would help keep things more easily organized

JWA111 commented on frontend/src/App.js:
```diff
@@ -0,0 +1,82 @@
+    setTodoItems(newToDoItems)
+  } 
+
+  const moveItem = (itemName,dragPosition) => {
```
probably should just name this `position` since this component doesn't have a concept of "drag"
```suggestion
  const moveItem = (itemName, position) => {
```

JWA111 commented on frontend/src/App.js:
```diff
@@ -0,0 +1,82 @@
+    return items.sort((a, b) => a.position - b.position);
+  }
+
+  const renameTodoItem = (oldItemName,newItemName)=>{
```
Continuing the discussion of having an `id` field, then it would make this renaming logic a little less messy. Also as it is currently implemented you cannot allow duplicates, but with an `id` we could allow for duplicates.

Currently, indexing by the text is causing some very buggy behavior

JWA111 commented on frontend/src/Components/CreateItemForm.js:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    const handleSubmit = (newToDoName)=>{
+        const trimmedToDo = newToDoName !== null && newToDoName !== undefined ? newToDoName.trim() : null;
```
Since JS uses truthy/falsey logic, this could be simplified to:

```suggestion
        const trimmedToDo = newToDoName && newToDoName.trim()
```

but either way all of this could be avoided by adding an early return to the start of the function like:
```js
if (!newToDoName)
  return
```

JWA111 commented on frontend/src/Components/CreateItemForm.js:
```diff
@@ -0,0 +1,41 @@
+        setNewToDoName(event.target.value);
+    }
+
+    const handleKeyPress = (key, newTodoName)=>{
+        if(key === 'Enter'){
+            handleSubmit(newTodoName);
+        }
+    }
```
pass the whole event to the function so that you can call `event.stopPropagation()` on it which will prevent it from propagating (continuing up) to parent elements

whats-a-handle replied:
@JWA111 Should I add `event.stopPropagation()` to other handlers like my `onClick` functions in other components too by default? Is there a good way to tell when I should include this or exclude this?

JWA111 commented on frontend/src/App.js:
```diff
@@ -0,0 +1,82 @@
+            <Grid container  >
+              <Grid item xs={2}/>
+              <Grid item xs={8} style={{backgroundColor:'white'}}>
+               <ToDoList markToDoItem={markToDoItem} renameTodoItem={renameTodoItem} deleteToDoItem={deleteToDoItem} moveItem={moveItem}>{todoItemsOrderedList}</ToDoList>
```
we should pass `todoItemsOrderedList` as a prop

JWA111 commented on frontend/src/Components/CreateItemForm.js:
```diff
@@ -0,0 +1,41 @@
+
+
+export default function CreateItemForm(props){
+    const [newToDoName,setNewToDoName] = useState('');
```
do we need this state?

JWA111 reviewed (changes requested):
This is a great start, here is some feedback for you to address:

To help with code style and formatting I'd recommend that you add [eslint](https://eslint.org/). The formatting is a bit all over the place and inconsistent.

It is good practice to define propTypes on your components to help the reader know what props the component takes.

I don't see much of your own CSS/style, it would be good to make your own style tweaks on top of the material-ui defaults. Some improvements you could make:

The white box which is really the main container of the list just grows with the height of the list. It could be improved by making the container a fixed height and making the list scrollable. Some slightly rounded corners and shadows might also make it look a little sleeker.
![image](https://user-images.githubusercontent.com/11657466/163632904-f4011aa1-128b-411c-a452-02bfa696b99f.png)

Move the button in-line with the input field: 
![image](https://user-images.githubusercontent.com/11657466/163633380-276e20ca-59d5-4e27-b35d-dfb0590b9b9f.png)

The input field is really close to the top, could use some padding
![image](https://user-images.githubusercontent.com/11657466/163634004-9b41e780-000e-4228-93a8-bc39dfaf504b.png)


When you have a lot of items, it becomes hard to distinguish which delete button goes with which item. It could use some faint bordering to show the row
![image](https://user-images.githubusercontent.com/11657466/163634078-5b6eaf09-7305-43e2-830c-2fa2ad308439.png)

whats-a-handle committed:
Add eslint and prop-types package and config

whats-a-handle committed:
Initial reformat and fix linting issues using airbnb style guide. Added PropTypes

whats-a-handle committed:
Added navbar with searchbar. Working on restyling todo items into cards

whats-a-handle committed:
Updated TodoCard with props and conditional cardaction styling

whats-a-handle committed:
Initial refactor of TodoList to use TodoCards

whats-a-handle committed:
Trying to fix heroku deploy errors with package version and missing TodoList component reference

whats-a-handle committed:
Removing TodoList to force cache update

whats-a-handle committed:
Added back todolist to fix cache issue on heroku

whats-a-handle committed:
remove unused NewCustomAppBar and conditionally hide search bar on xs screen size and replace with search icon

whats-a-handle committed:
Created 3 columns for Todo, Inprogress and Done instead of just one big todo list. Changed background color. Changed card sizing

whats-a-handle committed:
Moved column names inside columns and added more styling to columns

whats-a-handle committed:
More styling changes. Checkbox is now in card actions and only shows on mouseover when in inprogress or done columns. Resized and offset Done column. Columns buggy on mobile

whats-a-handle committed:
Added renderTodoCards method to simplify rendering of cards for columns in TodoList. Added minimum height for TodoList to prevent empty space when no items in columns

whats-a-handle committed:
Initial work on createitemform styling

whats-a-handle committed:
Formatting of TodoCard and list. TodoCard will display 5 lines of description and 2 line of title and will expand on mouseover. Grid structure nesting issue where other columns expand on mouseover...

whats-a-handle committed:
Initial refactor to move state to TodoList and out of App.jsx according to feedback. Working add item and delete item

whats-a-handle committed:
Add markTodoItem functionality

whats-a-handle committed:
Add sorting on TodoList and fix onClick function for markTodoItem on TodoCard

whats-a-handle committed:
Update todo title and description functionality

whats-a-handle committed:
Move inline styles to style objects for improved readability

whats-a-handle commented:
👋  Hey@JWA111, I'm ready for another round of review 👀 whenever you have some time. I reworked it using your feedback and added more of my own styling 😄 . I kept running into issues with the drag and drop functionality so that's currently not working - maybe we can pair on that sometime? Looking forward to more of your feedback! 

I also got it up on heroku https://wp-todo-list-app.herokuapp.com/

JWA111 commented on .vscode/settings.json:
```diff
@@ -0,0 +1,9 @@
+{
```
Committing this is fine, but for learning's sake I'll mention that by committing these settings you'd be forcing other contributors to use the same settings

whats-a-handle replied:
good to know! ill avoid committing this in the future 👍

JWA111 commented on frontend/src/Components/TodoList.jsx:
```diff
@@ -0,0 +1,134 @@
+    setTodoItems((prevItems) => ({ ...prevItems, [item.id]: { ...item } }));
+  };
+
+  React.useEffect(() => {
```
It would be better to import just the pieces you need rather than all of the react library. This is true of most libraries, and the reason is that some builders have a mechanism called "tree shaking" where only the pieces used will end up in the resulting code bundle. And by pulling in the entirety of the library it makes the build think you are using everything when it does the tree shaking.

whats-a-handle replied:
Makes sense, thanks 😄 ! I've fixed this now by replacing the import statement with `import { useState, useEffect } from 'react';` and then called `useEffect()` instead of `React.useEffect()`. 

I removed the unnecessary React imports in other files as well.

JWA111 commented on frontend/src/Components/TodoList.jsx:
```diff
@@ -0,0 +1,134 @@
+
+  const renderTodoCards = (items) => {
+    const todoItemsForSorting = [...items];
+    let todoCards = [];
+    sortItems(todoItemsForSorting);
+    if (todoItemsForSorting !== null && todoItemsForSorting !== undefined) {
+      todoCards = todoItemsForSorting.map((item) => (
```
If you wanted to use const instead of a reassignable `let` and keep the operation "safe" (not throwing an error when things are undefined) you could structure it like:

```suggestion
    const todoCards = (todoItemsForSorting || []).map((item) => (
```

JWA111 commented on frontend/src/Components/TodoList.jsx:
```diff
@@ -0,0 +1,134 @@
+  const renderTodoCards = (items) => {
+    const todoItemsForSorting = [...items];
+    let todoCards = [];
+    sortItems(todoItemsForSorting);
```
It is best to avoid "mutation" at all costs, as it can lead to buggy code. "Mutation" is when you modify a value rather that changing it into a new one. An object that you do not mutate is called "immutable". Writing immutable code is considered to be one of the best ways to reduce bugs in code.

I highly recommend that you read more about it: https://www.geeksforgeeks.org/why-is-immutability-so-important-in-javascript/

whats-a-handle replied:
thanks, that article was helpful 👍 . I've refactored this so that instead of mutating `todoItemsForSorting` with sort, I'm instead passing `items` in like so  `const sortedTodoItems = sortItems([...items]);` so now it won't mutate the original array that's getting passed in 😄

JWA111 reviewed (approved):
This is looking ready to move to part 2. I've mostly just found minor things that are good points of learning. I like that you've used spread operators and higher order functions, as many new devs miss those features. Feel free to fix up those things and merge the PR into the main branch. Then we can begin working on the backend.

whats-a-handle committed:
Remove unnecessary React import and update eslint to not require React import on every jsx file as well.

whats-a-handle committed:
Prevent unnecessary mutating of copy of item array when sorting and clean up undefined/null check logic
